types/yup/index.d.ts

```diff
@@ -105,39 +107,49 @@ export interface StringSchemaConstructor {
     new (): StringSchema;
 }
 
-export interface StringSchema extends Schema<string> {
-    length(limit: number | Ref, message?: TestOptionsMessage): StringSchema;
-    min(limit: number | Ref, message?: TestOptionsMessage): StringSchema;
-    max(limit: number | Ref, message?: TestOptionsMessage): StringSchema;
+export interface StringSchema<T = string> extends Schema<T> {
+    length(limit: number | Ref, message?: TestOptionsMessage): StringSchema<T>;
+    min(limit: number | Ref, message?: TestOptionsMessage): StringSchema<T>;
+    max(limit: number | Ref, message?: TestOptionsMessage): StringSchema<T>;
     matches(
         regex: RegExp,
         messageOrOptions?:
             | TestOptionsMessage
             | { message?: TestOptionsMessage; excludeEmptyString?: boolean }
-    ): StringSchema;
-    email(message?: TestOptionsMessage): StringSchema;
-    url(message?: TestOptionsMessage): StringSchema;
-    ensure(): StringSchema;
-    trim(message?: TestOptionsMessage): StringSchema;
-    lowercase(message?: TestOptionsMessage): StringSchema;
-    uppercase(message?: TestOptionsMessage): StringSchema;
+    ): StringSchema<T>;
+    email(message?: TestOptionsMessage): StringSchema<T>;
+    url(message?: TestOptionsMessage): StringSchema<T>;
+    ensure(): StringSchema<T>;
+    trim(message?: TestOptionsMessage): StringSchema<T>;
+    lowercase(message?: TestOptionsMessage): StringSchema<T>;
+    uppercase(message?: TestOptionsMessage): StringSchema<T>;
+    nullable(isNullable?: true): StringSchema<T | null>;
+    nullable(isNullable: false): StringSchema<Exclude<T, null>>;
 }
 
 export interface NumberSchemaConstructor {
     (): NumberSchema;
     new (): NumberSchema;
 }
 
-export interface NumberSchema extends Schema<number> {
-    min(limit: number | Ref, message?: TestOptionsMessage): NumberSchema;
-    max(limit: number | Ref, message?: TestOptionsMessage): NumberSchema;
-    lessThan(limit: number | Ref, message?: TestOptionsMessage): NumberSchema;
-    moreThan(limit: number | Ref, message?: TestOptionsMessage): NumberSchema;
-    positive(message?: TestOptionsMessage): NumberSchema;
-    negative(message?: TestOptionsMessage): NumberSchema;
-    integer(message?: TestOptionsMessage): NumberSchema;
-    truncate(): NumberSchema;
-    round(type: "floor" | "ceil" | "trunc" | "round"): NumberSchema;
+export interface NumberSchema<T = number> extends Schema<T> {
```

Comment:
Same thing here, I think we should change `T = number` to `T extends number` because the second way locks the client from passing in some other type into this that might not be a number altogether.

Reply:
👍